### PR TITLE
Minify scripts

### LINKED ISSUE #44: Allow for script minification
There should be a flag (`--minify` and `-m`) that causes uflash to "minify" the Python script using various heuristics so it'll fit into the limited amount of memory available.

See: https://github.com/mu-editor/mu/pull/385 for an example of what I mean.

### PR DESCRIPTION
**Pull request by ZanderBrown**
Partly resolves #44 but doesn't allow forced minification

Based on and supersedes work over in Mu

### CODE REVIEW

**Comment by carlosperate** [comment]
The nice thing about uFlash is that it works on it's own without having to install/pip install anything else. If the minification is going to be done in a separate module, could that be an optional dependency? So if it's not present uFlash could still work as it did.

File: uflash.py [hunk 1]
```diff
@@ -69,11 +71,13 @@ def hexlify(script):
     # Convert line endings in case the file was created on Windows.
     script = script.replace(b'\r\n', b'\n')
     script = script.replace(b'\r', b'\n')
+    if len(script) > _MAX_SIZE:
+        script = nudatus.mangle(script)
```
Comment (carlosperate) [inline uflash.py]: Looks like nudatus already contains the micro:bit internal flash size, but it could be used as a generic tool for any MicroPython code, so perhaps the size could be passed as an another argument?
This would also be helpful to use for `main.py` file flashed via filesystem, as it can be larger than 8KBs, but still has a limit.

Comment (ZanderBrown) [reply uflash.py]: I'm not aware of nudatus referencing the micro:bit flash size as I did intend for it to be independent, this len comparison is just to prevent unnecessary work

Makes sense as for nudatus to be an optional dependency though

Comment (carlosperate) [reply uflash.py]: Sorry, you are right, not quite sure why I got that impression. It's already generic as it is.

Comment (ZanderBrown) [reply uflash.py]: Don't worry, your probably remembering the original PR in Mu

File: requirements.txt [hunk 2]
```diff
@@ -4,6 +4,7 @@ pyflakes
 coverage
 sphinx
 pytest-cov
+nudatus>=0.0.2
```
Comment (ntoll) [inline requirements.txt]: Hmmm... so this breaks the "just copy the uflash script" for it to just work (tm) feature (important for use in schools where installing dependencies probably isn't an option). :-(

File: uflash.py [hunk 3]
```diff
@@ -21,6 +21,13 @@
 from subprocess import check_output
 import time
 
+# nudatus is an optional dependancy
+can_minify = True
+try:
+    import nudatus
+except ImportError:  # pragma: no cover
+    can_minify = False
```
Comment (ntoll) [inline uflash.py]: Ah, right... I see what you did there.

File: uflash.py [hunk 4]
```diff
@@ -69,11 +77,13 @@ def hexlify(script):
     # Convert line endings in case the file was created on Windows.
     script = script.replace(b'\r\n', b'\n')
     script = script.replace(b'\r', b'\n')
+    if can_minify and len(script) > _MAX_SIZE:
```
Comment (ntoll) [inline uflash.py]: As I mentioned over in the Mu repository, the minification should probably be behind a flag, so it's explicitly asked for. This feels a bit too much like magic, especially since the user may extract the script from the resulting hex file and get a surprise that comments / white-space etc are stripped.

**Comment by ntoll** [comment]
@ZanderBrown many thanks for this! I think the feature is important but I'm not sure we're there on the details yet. As you probably realise from my comments on the code, I'd like:

* To retain the "just copy the script" (tm) "feature" so there are no external dependencies.
* An explicit flag so the user is shown to *intend* to minify the script. Perhaps if the un-minified script is too big, there should be a suggestion to minify it in the message uflash outputs to the user.

As always, I'm all ears to alternatives to the above. :-)

**Comment by ntoll** [comment]
@ZanderBrown just read your comment in the Mu repository about splitting out the logic for minification should other projects want to re-use it. That's a really *good* point. Hmmm... not sure what to do WRT zero-dependencies uflash vs. re-use of nudatus. Could nudatus be merged into uflash..? Quite understand if you don't want to do this... although I'm quite happy to move uflash into a Mu-editor repository (so I cede "ownership") and update copyright to include you etc etc etc... Basically, I'm all ears for suggestions to resolve this. Perhaps add nudatus to Mu's contrib namespace (like uflash)..? Thoughts and ideas most welcome! I'm just thinking raw thoughts out loud... :-)

**Comment by ZanderBrown** [comment]
@ntoll some pretty raw thoughts in response:

> That's a really good point. Hmmm... not sure what to do WRT zero-dependencies uflash vs. re-use of nudatus. Could nudatus be merged into uflash..?

Hmm, the problem there is that uflash is micro:bit specific and i was hoping to be a little more generic, but equally uflash is quite small

> Quite understand if you don't want to do this... although I'm quite happy to move uflash into a Mu-editor repository (so I cede "ownership") and update copyright to include you etc etc etc...

Not the end of the world but having published to pypi ext it would nice to retain some kind of 'ownership' and I wouldn't want to start making a mess of the Mu commit log (or uflash for that matter)

Reminds me I should add a copyright notice to nudatus as it currently stands

> Basically, I'm all ears for suggestions to resolve this. Perhaps add nudatus to Mu's contrib namespace (like uflash)..? Thoughts and ideas most welcome! I'm just thinking raw thoughts out loud... :-)

For Mu obviously pip or the contrib namespace are perfectly fine as has been done with other dependencies, would defeat being open-source otherwise

uFlash is probably more complicated here as it's always tried to be a single all in one script and i don't want to 'break' that, which is why nudatus is an 'optional' import (following @carlosperate 's comment)

tl;dr I would like to keep nudatus as a project in a separate repo I control (although moving to `mu-editor` rather than `ZanderBrown` would be fine) but I'm not massively bothered how Mu/uFlash incorporate it barring the standard credit the author

**Comment by ntoll** [comment]
@ZanderBrown completely understand! Given your comments I think the solution is:

* Keep nudatus separate from uFlash for all the good reasons you state.
* I'd prefer to put nudatus into `contrib` simply because it'll make Debian packaging less of a PiA due to contrib-y things not being in Debian (yet).
* We retain @carlosperate's suggested changes for checking for nudatus. In any case, I'll add nudatus to `setup.py` so anyone using pip gets to be able to minify. ;-)

I'll merge this tomorrow since I'm off on Dad taxi duties 10 minutes ago... ;-) (yeah I'm late).

**Comment by ZanderBrown** [comment]
:tada: Nice to see it come together
